packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -270,10 +270,19 @@ abstract class ImageProvider<T> {
   /// This is the public entry-point of the [ImageProvider] class hierarchy.
   ///
   /// Subclasses should implement [obtainKey] and [load], which are used by this
-  /// method.
+  /// method. If they need to manage the actual resolution of the image, they
+  /// should override [resolveForStream] instead of this method.
+  @nonVirtual
```

Comment:
This is a break.

Reply:
why must it be nonVirtual? I mean, I get that it's useful for catching people who are using the API the old way, but is it strictly necessary? What if someone wants to take an entirely different approach?

Reply:
Because if some other provider overrides it, its new logic will never get called by the new kind of provider I'm making that would wrap it.

I imagine other providers would have this problem. Plus the logic in it is very complicated.

Alternatively, we could update docs to say "if you override this, composed providers may not work and you should consider overriding useKey instead" - but I'm inclined to think people are more likely to see an analyzer warning than a doc comment.